File: src/test/java/net/pterodactylus/sone/freenet/wot/IdentityChangeEventSenderTest.java
/*
 * Sone - IdentityChangeEventSenderTest.java - Copyright © 2013–2016 David Roden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.pterodactylus.sone.freenet.wot;

import static com.google.common.collect.ImmutableMap.of;
import static java.util.Arrays.asList;
import static net.pterodactylus.sone.freenet.wot.Identities.createIdentity;
import static net.pterodactylus.sone.freenet.wot.Identities.createOwnIdentity;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;

import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import net.pterodactylus.sone.freenet.wot.event.IdentityAddedEvent;
import net.pterodactylus.sone.freenet.wot.event.IdentityRemovedEvent;
import net.pterodactylus.sone.freenet.wot.event.IdentityUpdatedEvent;
import net.pterodactylus.sone.freenet.wot.event.OwnIdentityAddedEvent;
import net.pterodactylus.sone.freenet.wot.event.OwnIdentityRemovedEvent;

import com.google.common.eventbus.EventBus;
import org.junit.Test;

/**
 * Unit test for {@link IdentityChangeEventSender}.
 */
public class IdentityChangeEventSenderTest {

	private final EventBus eventBus = mock(EventBus.class);
	private final List<OwnIdentity> ownIdentities = asList(
			createOwnIdentity("O1", asList("Test"), of("KeyA", "ValueA")),
			createOwnIdentity("O2", asList("Test2"), of("KeyB", "ValueB")),
			createOwnIdentity("O3", asList("Test3"), of("KeyC", "ValueC"))
	);
	private final List<Identity> identities = asList(
			createIdentity("I1", Collections.<String>emptyList(), Collections.<String, String>emptyMap()),
			createIdentity("I2", Collections.<String>emptyList(), Collections.<String, String>emptyMap()),
			createIdentity("I3", Collections.<String>emptyList(), Collections.<String, String>emptyMap()),
			createIdentity("I2", asList("Test"), Collections.<String, String>emptyMap())
	);
	private final IdentityChangeEventSender identityChangeEventSender = new IdentityChangeEventSender(eventBus, createOldIdentities());

	@Test
	public void addingAnOwnIdentityIsDetectedAndReportedCorrectly() {
		Map<OwnIdentity, Collection<Identity>> newIdentities = createNewIdentities();
		identityChangeEventSender.detectChanges(newIdentities);
		verify(eventBus).post(eq(new OwnIdentityRemovedEvent(ownIdentities.get(0))));
		verify(eventBus).post(eq(new IdentityRemovedEvent(ownIdentities.get(0), identities.get(0))));
		verify(eventBus).post(eq(new IdentityRemovedEvent(ownIdentities.get(0), identities.get(1))));
		verify(eventBus).post(eq(new OwnIdentityAddedEvent(ownIdentities.get(2))));
		verify(eventBus).post(eq(new IdentityAddedEvent(ownIdentities.get(2), identities.get(1))));
		verify(eventBus).post(eq(new IdentityAddedEvent(ownIdentities.get(2), identities.get(2))));
		verify(eventBus).post(eq(new IdentityRemovedEvent(ownIdentities.get(1), identities.get(0))));
		verify(eventBus).post(eq(new IdentityAddedEvent(ownIdentities.get(1), identities.get(2))));
		verify(eventBus).post(eq(new IdentityUpdatedEvent(ownIdentities.get(1), identities.get(1))));
	}

	private Map<OwnIdentity, Collection<Identity>> createNewIdentities() {
		Map<OwnIdentity, Collection<Identity>> oldIdentities = new HashMap<OwnIdentity, Collection<Identity>>();
		oldIdentities.put(ownIdentities.get(1), asList(identities.get(3), identities.get(2)));
		oldIdentities.put(ownIdentities.get(2), asList(identities.get(1), identities.get(2)));
		return oldIdentities;
	}

	private Map<OwnIdentity, Collection<Identity>> createOldIdentities() {
		Map<OwnIdentity, Collection<Identity>> oldIdentities = new HashMap<OwnIdentity, Collection<Identity>>();
		oldIdentities.put(ownIdentities.get(0), asList(identities.get(0), identities.get(1)));
		oldIdentities.put(ownIdentities.get(1), asList(identities.get(0), identities.get(1)));
		return oldIdentities;
	}

}

File: src/test/kotlin/net/pterodactylus/sone/freenet/wot/IdentityChangeEventSenderTest.kt
/*
 * Sone - IdentityChangeEventSenderTest.kt - Copyright © 2013–2020 David Roden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package net.pterodactylus.sone.freenet.wot

import com.google.common.eventbus.EventBus
import com.google.common.eventbus.Subscribe
import net.pterodactylus.sone.freenet.wot.event.IdentityAddedEvent
import net.pterodactylus.sone.freenet.wot.event.IdentityRemovedEvent
import net.pterodactylus.sone.freenet.wot.event.IdentityUpdatedEvent
import net.pterodactylus.sone.freenet.wot.event.OwnIdentityAddedEvent
import net.pterodactylus.sone.freenet.wot.event.OwnIdentityRemovedEvent
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.contains
import org.hamcrest.Matchers.containsInAnyOrder
import org.junit.Test

/**
 * Unit test for [IdentityChangeEventSender].
 */
class IdentityChangeEventSenderTest {

	private val eventBus = EventBus()
	private val ownIdentities = listOf(
			createOwnIdentity("O1", setOf("Test"), "KeyA" to "ValueA"),
			createOwnIdentity("O2", setOf("Test2"), "KeyB" to "ValueB"),
			createOwnIdentity("O3", setOf("Test3"), "KeyC" to "ValueC")
	)
	private val identities = listOf(
			createIdentity("I1", setOf()),
			createIdentity("I2", setOf()),
			createIdentity("I3", setOf()),
			createIdentity("I2", setOf("Test"))
	)
	private val identityChangeEventSender = IdentityChangeEventSender(eventBus, createOldIdentities())

	@Test
	fun addingAnOwnIdentityIsDetectedAndReportedCorrectly() {
		val newIdentities = createNewIdentities()
		val ownIdentityRemovedEvents = mutableListOf<OwnIdentityRemovedEvent>()
		eventBus.register(object {
			@Subscribe
			fun ownIdentityRemoved(ownIdentityRemovedEvent: OwnIdentityRemovedEvent) {
				ownIdentityRemovedEvents += ownIdentityRemovedEvent
			}
		})
		val identityRemovedEvents = mutableListOf<IdentityRemovedEvent>()
		eventBus.register(object {
			@Subscribe
			fun identityRemovedEvent(identityRemovedEvent: IdentityRemovedEvent) {
				identityRemovedEvents += identityRemovedEvent
			}
		})
		val owIdentityAddedEvents = mutableListOf<OwnIdentityAddedEvent>()
		eventBus.register(object {
			@Subscribe
			fun ownIdentityAdded(ownIdentityAddedEvent: OwnIdentityAddedEvent) {
				owIdentityAddedEvents += ownIdentityAddedEvent
			}
		})
		val identityAddedEvents = mutableListOf<IdentityAddedEvent>()
		eventBus.register(object {
			@Subscribe
			fun identityAdded(identityAddedEvent: IdentityAddedEvent) {
				identityAddedEvents += identityAddedEvent
			}
		})
		val identityUpdatedEvents = mutableListOf<IdentityUpdatedEvent>()
		eventBus.register(object {
			@Subscribe
			fun identityUpdatedEvent(identityUpdatedEvent: IdentityUpdatedEvent) {
				identityUpdatedEvents += identityUpdatedEvent
			}
		})
		identityChangeEventSender.detectChanges(newIdentities)
		assertThat(ownIdentityRemovedEvents, contains(OwnIdentityRemovedEvent(ownIdentities[0])))
		assertThat(identityRemovedEvents, containsInAnyOrder(
				IdentityRemovedEvent(ownIdentities[0], identities[0]),
				IdentityRemovedEvent(ownIdentities[0], identities[1]),
				IdentityRemovedEvent(ownIdentities[1], identities[0])
		))
		assertThat(owIdentityAddedEvents, contains(OwnIdentityAddedEvent(ownIdentities[2])))
		assertThat(identityAddedEvents, containsInAnyOrder(
				IdentityAddedEvent(ownIdentities[2], identities[1]),
				IdentityAddedEvent(ownIdentities[2], identities[2]),
				IdentityAddedEvent(ownIdentities[1], identities[2])
		))
		assertThat(identityUpdatedEvents, contains(IdentityUpdatedEvent(ownIdentities[1], identities[1])))
	}

	private fun createNewIdentities() = mapOf(
			ownIdentities[1] to listOf(identities[3], identities[2]),
			ownIdentities[2] to listOf(identities[1], identities[2])
	)

	private fun createOldIdentities() = mapOf(
			ownIdentities[0] to listOf(identities[0], identities[1]),
			ownIdentities[1] to listOf(identities[0], identities[1])
	)

}
